File: LeetCode/Unknown/백준/Gold/15681. 트리와 쿼리/트리와 쿼리.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

    static int N, R, Q;
    static List<List<Integer>> graph;
    static int[] subTree;
    static boolean[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        R = Integer.parseInt(st.nextToken());
        Q = Integer.parseInt(st.nextToken());

        graph = new ArrayList<>();
        for (int i = 0; i <= N; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < N - 1; i++) {
            st = new StringTokenizer(br.readLine());
            int U = Integer.parseInt(st.nextToken());
            int V = Integer.parseInt(st.nextToken());

            graph.get(U).add(V);
            graph.get(V).add(U);
        }

        visited = new boolean[N + 1];
        subTree = new int[N + 1];

        makeTree(R);

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < Q; i++) {
            int q = Integer.parseInt(br.readLine());
            sb.append(subTree[q]).append('\n');
        }

        System.out.print(sb);
    }

    private static int makeTree(int cur) {
        visited[cur] = true;
        int count = 1;

        for (int next : graph.get(cur)) {
            if (!visited[next]) {
                count += makeTree(next);
            }
        }

        subTree[cur] = count;
        return count;
    }
}

File: LeetCode/Unknown/백준/Gold/15681. 트리와 쿼리/트리와 쿼리.cc
#include <iostream>
#include <vector>

using namespace std;

vector<vector<int>> graph;
vector<int> subtree;
vector<bool> visited;

int dfs(int node) {
    visited[node] = true;
    int count = 1;

    for (int next : graph[node]) {
        if (!visited[next]) count += dfs(next);
    }

    return subtree[node] = count;
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    int N, R, Q;
    cin >> N >> R >> Q;

    graph.resize(N + 1);
    subtree.resize(N + 1);
    visited.resize(N + 1, false);

    for (int i = 0; i < N - 1; i++) {
        int u, v;
        cin >> u >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }

    dfs(R);

    while (Q--) {
        int node;
        cin >> node;
        cout << subtree[node] << "\n";
    }

    return 0;
}
